wire/msgtx.go

```diff
@@ -336,9 +336,10 @@ func (msg *MsgTx) TxHash() chainhash.Hash {
 	// Ignore the error returns since the only way the encode could fail
 	// is being out of memory or due to nil pointers, both of which would
 	// cause a run-time panic.
-	buf := bytes.NewBuffer(make([]byte, 0, msg.SerializeSizeStripped()))
-	_ = msg.SerializeNoWitness(buf)
-	return chainhash.DoubleHashH(buf.Bytes())
+	txHash := sha256.New()
+	_ = msg.SerializeNoWitness(txHash)
+	bytes := chainhash.DoubleHashRaw(txHash)
+	return *((*[32]byte)(bytes))
```

Comment:
Funk casting and dereferencing here...

Could use something like this maybe? https://go.dev/play/p/cO7KIylsYA2

Unsure re diff in efficiency though.

Reply:
It's using the slice -> array casting made available in go1.17.

https://stackoverflow.com/questions/30285680/how-to-convert-slice-to-fixed-size-array


It can avoid the costly `copy()` so I'm sure that function would have a significant penalty wrt memory allocation. I'll write some benchmarks comparing the two and we could switch over if it doesn't have that big of a performance penalty.

Reply:
So the performance seems to be about the same so either would be ok. Here's the benchstat results:

```
goos: linux
goarch: amd64
pkg: github.com/btcsuite/btcd/wire
cpu: AMD Ryzen 5 3600 6-Core Processor              
          │ new-multitx.txt │      bytes-to-array.txt       │
          │     sec/op      │   sec/op     vs base          │
TxHash-10       1.543µ ± 2%   1.548µ ± 0%  ~ (p=0.196 n=10)

          │ new-multitx.txt │       bytes-to-array.txt       │
          │      B/op       │    B/op     vs base            │
TxHash-10        160.0 ± 0%   160.0 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

          │ new-multitx.txt │       bytes-to-array.txt       │
          │    allocs/op    │ allocs/op   vs base            │
TxHash-10        2.000 ± 0%   2.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

But if we use generics, we'd have to bump up the go minimum version to 1.18. Here's something different that's easier to read and only requires go1.17:

```
// TxHash generates the Hash for the transaction.
func (msg *MsgTx) TxHash() chainhash.Hash {
        // Encode the transaction and calculate double sha256 on the result.
        // Ignore the error returns since the only way the encode could fail
        // is being out of memory or due to nil pointers, both of which would
        // cause a run-time panic.
        txHash := sha256.New()
        _ = msg.SerializeNoWitness(txHash)
        res := (*[32]byte)(chainhash.DoubleHashRaw(txHash))
        return *res
}
```

Which one looks better?

Reply:
Yeah re generics, we found out that some people depend on the project for `gopherjs` transliteration. Moving to add generics now would break those projects....

I don't think we should hold of indefinitely though, as transliteration if pretty fragile in any case.